precompiles/src/precompile_set.rs

```diff
@@ -375,19 +361,23 @@ fn common_checks<R: pallet_evm::Config, C: PrecompileChecks>(
 		u32::from_be_bytes(buffer)
 	});
 
-	// Is this selector callable from a smart contract?
-	let callable_by_smart_contract =
-		C::callable_by_smart_contract(caller, selector).unwrap_or(false);
-	if !callable_by_smart_contract && !is_address_eoa_or_precompile::<R>(handle, caller)? {
-		return Err(revert("Function not callable by smart contracts"));
-	}
+	let caller_address_type = get_address_type::<R>(handle, caller)?;
```

Comment:
nit?

```suggestion
	let caller_address_type = get_address_type::<R>(handle, caller)?;
	
	match caller_address_type {
    AddressType::Precompile => {
        let callable_by_precompile = C::callable_by_precompile(caller, selector).unwrap_or(false);
        if !callable_by_precompile {
            return Err(revert("Function not callable by precompiles"));
        }
    },
    AddressType::Contract => {
        let callable_by_smart_contract = 
            C::callable_by_smart_contract(caller, selector).unwrap_or(false);
        if !callable_by_smart_contract {
            return Err(revert("Function not callable by smart contracts"));
        }
    },
    _ => {} // Other address types (like EOA) don't need specific checks
}
```

Reply:
I like the suggestion, added in 57a205d08342e8c226c7a26a0320d18914c971a3